databricks_cli/stack/api.py

```diff
@@ -234,6 +245,58 @@ def _update_job(self, job_settings, job_id):
         click.echo("Updating Job")
         self.jobs_client.reset_job({'job_id': job_id, 'new_settings': job_settings})
 
+    def _deploy_workspace(self, resource_properties, physical_id, overwrite):
+        """
+        Deploy workspace asset.
+
+        :param resource_properties: dict of properties for the workspace asset. Must contain the
+        'source_path' and 'path' fields. The other fields will be inferred if not provided.
+        :param physical_id: dict containing physical identifier of workspace asset on databricks.
+        Should contain the field 'path'.
+        :param overwrite: Whether or not to overwrite the contents of workspace notebooks.
+        :return: (dict, dict) of (physical_id, deploy_output). physical_id is the physical ID for
+        the stack status that contains the workspace path of the notebook or directory on datbricks.
+        deploy_output. Is the initial information about the asset on databricks at deploy time.
+        """
+        try:
+            local_path = resource_properties['source_path']
+            workspace_path = resource_properties['path']
+        except KeyError as e:
+            raise StackError("%s doesn't exist in workspace resource properties" % str(e))
+
+        lang_fmt = WorkspaceLanguage.to_language_and_format(local_path)  # Guess language and format
+        language, fmt = None, None
+        if lang_fmt:
+            language, fmt = lang_fmt
+
+        # Plug in optional properties.
+        if 'language' in resource_properties:
+            language = resource_properties['language']
+        if 'format' in resource_properties:
+            fmt = resource_properties['format']
+
+        object_type = "DIRECTORY" if os.path.isdir(local_path) else "NOTEBOOK"
+        if 'object_type' in resource_properties:
+            object_type = resource_properties['object_type']
+
```

Comment:
The top 15 lines handle the optional properties and additional inference on them.